src/number_theory/zeta_function.lean

```diff
@@ -0,0 +1,347 @@
+
+* `riemann_zeta`: the Riemann zeta function `ζ : ℂ → ℂ`.
+* `riemann_completed_zeta`: the completed zeta function `Λ : ℂ → ℂ`, which satisfies
+  `Λ(s) = π ^ (-s / 2) Γ(s / 2) ζ(s)` (away from the poles of `Γ(s / 2)`).
+* `riemann_completed_zeta₀`: the entire function `Λ₀` satisfying
+  `Λ₀(s) = Λ(s) + 1 / (s - 1) - 1 / s` wherever the RHS is defined.
+
+Note that mathematically `ζ(s)` is undefined at `s = 1`, while `Λ(s)` is undefined at both `s = 0`
+and `s = 1`. Our construction assigns some values at these points (which are not arbitrary, but
+I haven't checked exactly what they are).
+
+## Main results:
+
+* `differentiable_completed_zeta₀` : the function `Λ₀(s)` is entire.
+* `differentiable_at_completed_zeta`: the function `Λ(s)` is differentiable away from `s = 0` and
+  `s = 1`.
+
+## Outline of proofs:
+
+We define two related functions on the reals, `zeta_kernel₁` and `zeta_kernel₂`. The first is
+`(Θ (t * I) - 1) / 2`, where `θ` is Jacobi's theta function; its Mellin transform is exactly the
+completed zeta function. The second is obtained by subtracting a linear combination of powers on
+the interval `Ioc 0 1` to give a function with exponential decay at both `0` and `∞`. We then define
+`riemann_completed_zeta₀` as the Mellin transform of the second zeta kernel, and define
+`riemann_completed_zeta` and `riemann_zeta` from this.
+-/
+
+open measure_theory set filter asymptotics topological_space real asymptotics
+open complex (hiding exp norm_eq_abs abs_of_nonneg abs_two continuous_exp)
+
+open_locale topology real
+
+noncomputable theory
+
+/-!
+## Definition of the Riemann zeta function and related functions
+-/
+
+/-- Function whose Mellin transform is `π ^ (-s) * Γ(s) * zeta (2 * s)`, for `1 / 2 < Re s`. -/
+def zeta_kernel₁ (t : ℝ) : ℂ := ∑' (n : ℕ), rexp (-π * t * (n + 1) ^ 2)
+
+/-- Modified zeta kernel, whose Mellin transform is entire. --/
+def zeta_kernel₂ : ℝ → ℂ  := zeta_kernel₁ + indicator (Ioc 0 1) (λ t, (1 - 1 / sqrt t) / 2)
+
+/-- The completed Riemann zeta function with its poles removed, `Λ(s) + 1 / s - 1 / (s - 1)`. -/
+def riemann_completed_zeta₀ (s : ℂ) : ℂ := mellin zeta_kernel₂ (s / 2)
+
+/-- The completed Riemann zeta function, `Λ(s)`, which satisfies
+`Λ(s) = π ^ (-s / 2) Γ(s / 2) ζ(s)` (up to a minor correction at `s = 0`). -/
+def riemann_completed_zeta (s : ℂ) : ℂ := riemann_completed_zeta₀ s - 1 / s + 1 / (s - 1)
+
+/-- The Riemann zeta function `ζ(s)`. -/
+def riemann_zeta := function.update
+  (λ s : ℂ, ↑π ^ (s / 2) * riemann_completed_zeta s / Gamma (s / 2)) 0 (-1 / 2)
+
+/- Note the next lemma is true by definition; what's hard is to show that with this definition, `ζ`
+is continuous (and indeed analytic) at 0.
+
+TODO: Prove this -- this will be in a future PR. -/
+/-- We have `ζ(0) = -1 / 2`. -/
+lemma riemann_zeta_zero : riemann_zeta 0 = -1 / 2 := function.update_same _ _ _
+
+/- Set `riemann_zeta` to be irreducible to hide messy implementation details -/
+attribute [irreducible] riemann_zeta
```

Comment:
This attribute should be put directly on the definition, as in Lean 4 it is not possible to add the attribute after the fact. The only difference it makes is that you will need to unfold the definition in the proof of `riemann_zeta_zero`.